compiler/testData/codegen/box/diagnostics/functions/tailRecursion/defaultArgsWithSideEffects2.kt

```diff
@@ -0,0 +1,33 @@
+// DONT_RUN_GENERATED_CODE: JS
+
+var counter = 0
+fun inc() = counter++
+
+tailrec fun test(x: Int = 0, y: Int = inc(), z: Int = inc()) {
+    if (x * 2 != y || z - y != 1)
+        throw IllegalArgumentException()
+
+    if (x < 100000)
+        test(x + 1, y + 2, z + 2)
```

Comment:
Please switch this one to  `test(x + 1,  z = counter + 1)`

Reply:
test(z = sideEffect, x = sideEffect())

Reply:
Did you mean `test(y = inc(), x = x + 1)`? The check expects that:
1. `counter` is approximately `2 * x`. So `x = inc()` makes the the test unhappy.
2. `y = inc()` happens before `z = inc()`. So we cannot write `z = sideEffect()` without specifying `y`.

I'm uploading `test(y = inc(), x = x + 1)` for now.

Reply:
Actually I want to test argument initializion order in mixed case of named arguments and default values.
Please change this test to 
```
// DONT_RUN_GENERATED_CODE: JS

var counter = 0
fun inc() = counter++

tailrec fun test(x: Int = 0, y: Int = inc(), z: Int = inc()) {
    if (x * 2 != y + 1 || z - y != -1)
        throw IllegalArgumentException("x=$x y=$y z=$z")

    if (x < 100000)
        test(z = inc(), x = (counter - 1)/2 + 1)
}

fun box() : String {
    counter = 4
    test(1, 1, 0)

    counter = 0
    test(-1, -3, -4)

    return "OK"
}
```

Reply:
Done. Thanks for elaborating this!